### PR TITLE
Merge PR #2 and #3 and add compatibility with Theano and CNTK backend

### PR DESCRIPTION
**Pull request by drauh**
* This PR is a merge of my last two PR (#2 and #3)
* add compatibility with Theano and CNTK

### CODE REVIEW

**Comment by drauh** [comment]
I ran a small test overnight on the ptb dataset with the small configuration, --aug and --tying options. Results after 15 epochs are :
* val_acc: 0.2045
* val_perplexity: 193.1504

**Comment by icoxfog417** [comment]
Thank you for your pull request! Could you let me some check?

### About TimeDistributed

You remove the `TimeDistributed` in the `one_hot_model.py` but it is necessary for [many-to-many(2) type network](https://stackoverflow.com/questions/42755820/how-to-use-return-sequences-option-and-timedistributed-layer-in-keras).  

image

It is different from the ordinary language model implementation.

Ordinary implementation of the language model is as follows (Chainer/TensorFlow/PyTorch etc). To implements this, we have to use stateful LSTM, and it is difficult (I tried to do this and [you can see its result](https://github.com/icoxfog417/tying-wv-and-wc#experiments)).

image

([PyTorch code is good to understand this model](https://github.com/pytorch/examples/blob/master/word_language_model/main.py#L128))

### About cross_entropy order

In the `K.backend`, the argument order is `pred`, `true`. You can confirm it at [here](https://github.com/fchollet/keras/blob/2.0.6/keras/losses.py#L49).

But it is so confusable so [the fix is added recently](https://github.com/fchollet/keras/commit/788c56b9c60120e42f5d6cbd269499b878ba4859#diff-0b67ba4d54b9e4cff5dab56adc9cd160). But it is not deployed to PyPI (will be included 2.0.7).

I want to fix this order so I'll merge your request after above fix is available.

`PerplexityLogger` is a good idea! I want to use it in another project.

**Comment by drauh** [comment]
## About TimeDistributed

I thought that TimeDistributed(Dense()) was needed when keras Dense layer could only be applied to rank 2 tensor. So that to my understanding when applying to a rank 3 tensor :   
* Dense == Conv1D(kernel_size=1) = TimeDistrbuted(Dense())

And that's what the said [here](https://stackoverflow.com/questions/42755820/how-to-use-return-sequences-option-and-timedistributed-layer-in-keras#comment72635994_42758532). However I don't understand the interaction with statefulness of LSTM.

## About cross_entropy order

Good catch! I was using the master version and didn't pay enough attention...

## About stateful LSTMM

Because you used sequence size of 20, it means that you are doing BPTT up to 20 timesteps which seems enough. But this help to better initialize the LSTM cell state.

Some [research](https://arxiv.org/abs/1612.08083) actually use Conv1D to do language modeling, which is faster because easily parallelizable. And it is compatible with weight tying and augmented loss, so I want to test it when I have the time.
